storage/s3.go

```diff
@@ -32,6 +32,11 @@ var _ Storage = (*S3)(nil)
 
 var sentinelURL = urlpkg.URL{}
 
+var (
+	mutex            = &sync.Mutex{}
```

Comment:
If you use sync.Mutex for session management, read operations will block each other. We should allow multiple goroutines to access same session at the same time. Please try to use [sync.RWMutex ](https://golang.org/pkg/sync/#RWMutex)

Reply:
hmm, here if session has already been created it immediately returns. if we use rwmutex, two different goroutines might read the map at the 'same' time and if the map doesn't contain session, each will try to create new session one after the other

Reply:
👍